### PR TITLE
Adding support for neo4j

### PR DESCRIPTION
**Pull request by kaiserbergin**
This PR should add support for neo4j to test containers, as discussed in discussion #579

Some notes:
- neo4j is the default db name, I don't believe there are env variables to auto create a different one
- using a different username throws an error by the db, hence sticking it with `neo4j`
- docker image also expects NEO4J_AUTH as a combo of username and password, so I've tried to set that up in a way that makes sense

### CODE REVIEW

**Comment by HofmeisterAn** [comment]
Thanks for your pull request. I need a few more days until I can do the review (I am on the road).

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+﻿namespace DotNet.Testcontainers.Configurations
+{
+  using System;
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs:3]: Remove unnecessary `using`.

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    /// Initializes a new instance of the <see cref="Neo4jTestcontainerConfiguration" /> class.
+    /// </summary>
+    /// <param name="image">The Docker image.</param>
+    public Neo4jTestcontainerConfiguration(string image)
+      : base(image, BoltPort)
+    {
+    }
+
+    /// <inheritdoc />
+    public override string Database => DefaultDatabase;
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs]: If database is a const, I suggest something like:

```csharp
get => "neo4j";
set => throw new NotImplementedException();
```

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    /// <inheritdoc />
+    public override string Username => DefaultUsername;
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs]: If username is a const, I suggest something like:

```csharp
get => "neo4j";
set => throw new NotImplementedException();
```

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    /// <inheritdoc />
+    public override string Password
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs:49]: Maybe something like:

```csharp
get => this.Environments["NEO4J_AUTH"].Split(new[] { '/' }, 2, StringSplitOptions.RemoveEmptyEntries).Last();
set => this.Environments["NEO4J_AUTH"] = string.Join("/", this.Username, value);
```

and set the default password in the constructor? Then you can get ride of all default consts.

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    private const string DefaultPassword = "connect";
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Neo4jTestcontainerConfiguration" /> class.
+    /// </summary>
+    public Neo4jTestcontainerConfiguration()
+      : this(Neo4jImage)
+    {
+      this.Environments["NEO4JLABS_PLUGINS"] = "[\"apoc\"]";
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs]: Move both lines (`this.Environments[...]`) to the constructor below.

File: tests/Testcontainers.Tests/Fixtures/Containers/Unix/Modules/Databases/Neo4jFixture.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  using DotNet.Testcontainers.Builders;
+  using DotNet.Testcontainers.Configurations;
+  using DotNet.Testcontainers.Containers;
+  using JetBrains.Annotations;
+  using Neo4j.Driver;
+
+  [UsedImplicitly]
+  public sealed class Neo4jFixture : DatabaseFixture<Neo4jTestcontainer, IDriver>
+  {
+    private readonly TestcontainerDatabaseConfiguration configuration = new Neo4jTestcontainerConfiguration { Username = "username", Password = "password" };
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Tests/Fixtures/Containers/Unix/Modules/Databases/Neo4jFixture.cs]: It does not make sense to set the `Username` property based on your implementation.

File: tests/Testcontainers.Tests/Fixtures/Containers/Unix/Modules/Databases/Neo4jFixture.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        .ConfigureAwait(false);
+
+      this.Connection = GraphDatabase.Driver(
+        this.Container.ConnectionString,
+        AuthTokens.Basic(this.Container.Username, this.Container.Password));
+    }
+
+    public override async Task DisposeAsync()
+    {
+      await this.Container.DisposeAsync()
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Tests/Fixtures/Containers/Unix/Modules/Databases/Neo4jFixture.cs:34]: Dispose connection too (`this.Connection.Dispose();`).

File: tests/Testcontainers.Tests/Unit/Containers/Unix/Modules/Databases/Neo4jTestcontainerTest.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+    public Neo4jTestcontainerTest(Neo4jFixture neo4jFixture)
+    {
+      this.neo4jFixture = neo4jFixture;
+    }
+
+    [Fact]
+    public async Task ConnectionEstablished()
+    {
+      // Given
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Tests/Unit/Containers/Unix/Modules/Databases/Neo4jTestcontainerTest.cs]:
```csharp
var exception = await Record.ExceptionAsync(() => this.neo4jFixture.Connection.VerifyConnectivityAsync())
  .ConfigureAwait(false);

Assert.Null(exception);
```

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thank you for your contribution. I have added a few suggestions for improvement.

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs (hunk 1 above)
Comment (kaiserbergin) [reply src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs:3]: It's now used due to the updates requested.

**Comment by kaiserbergin** [comment]
@HofmeisterAn I've pushed up what I believe to be the changes you intended. Let me know if everything looks all right, or if you'd like any changes. Sorry it took me a bit to get back to you on this one, I should have a quicker turn around moving forward.

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  using System;
+  using System.Linq;
+  using DotNet.Testcontainers.Builders;
+  using JetBrains.Annotations;
+
+  /// <inheritdoc cref="TestcontainerDatabaseConfiguration" />
+  [PublicAPI]
+  public class Neo4jTestcontainerConfiguration : TestcontainerDatabaseConfiguration
+  {
+    private const string Neo4jImage = "neo4j:latest";
```
Comment (HofmeisterAn) [inline src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs]: Sorry, looks like I missed that. Can you pin the version please?

**Review by HofmeisterAn — APPROVED** [review]
> Sorry it took me a bit to get back to you on this one, I should have a quicker turn around moving forward.

No, worries. As you have time, there is no need to hurry. Can you pin the version, then I am happy to merge it. Thanks for your contribution.

File: src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs (hunk 9 above)
Comment (kaiserbergin) [reply src/Testcontainers/Configurations/Modules/Databases/Neo4jTestcontainerConfiguration.cs]: you got it, thanks!

**Review by HofmeisterAn — APPROVED** [review]

**Comment by HofmeisterAn** [comment]
9ab251cd2ace82a64d9a9cec046f6363400b2043, closes #579.

**Pull request closed without merging**